### PR TITLE
Distinguish invocation and pattern substitutions on SILFunctionType

### PR DESCRIPTION
**Pull request by rjmccall**
In order to allow this, I've had to rework the syntax of substituted function types; what was previously spelled `<T> in () -> T for <X>` is now spelled `@substituted <T> () -> T for <X>`.  I think this is a nice improvement for readability, but it did require me to churn a lot of test cases.

Distinguishing the substitutions has two chief advantages over the existing representation.  First, the semantics seem quite a bit clearer at use points; the `implicit` bit was very subtle and not always obvious how to use.  More importantly, it allows the expression of generic function types that must satisfy a particular generic abstraction pattern, which was otherwise impossible to express.

As an example of the latter, consider the following protocol conformance:

```
protocol P { func foo() }
struct A<T> : P { func foo() {} }
```

The lowered signature of `P.foo` is `<Self: P> (@in_guaranteed Self) -> ()`.  Without this change, the lowered signature of `A.foo`'s witness would be `<T> (@in_guaranteed A<T>) -> ()`, which does not preserve information about the conformance substitution in any useful way.  With this change, the lowered signature of this witness could be `<T> @Substituted <Self: P> (@in_guaranteed Self) -> () for <A<T>>`, which nicely preserves the exact substitutions which relate the witness to the requirement.

When we adopt this, it will both obviate the need for the special witness-table conformance field in SILFunctionType and make it far simpler for the SILOptimizer to devirtualize witness methods.  This patch does not actually take that step, however; it merely makes it possible to do so.

As another piece of unfinished business, while `SILFunctionType::substGenericArgs()` conceptually ought to simply set the given substitutions as the invocation substitutions, that would disturb a number of places that expect that method to produce an unsubstituted type.  This patch only set invocation arguments when the generic type is a substituted type, which we currently never produce in type-lowering.

My plan is to start by producing substituted function types for accessors.  Accessors are an important case because the coroutine continuation function is essentially an implicit component of the function type which the current substitution rules simply erase the intended abstraction of.  They're also used in narrower ways that should exercise less of the optimizer.

### CODE REVIEW

**Comment by rjmccall** [comment]
@swift-ci Please test

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by varungandhi-apple** [comment]
Does `docs/SIL.rst` need to be updated to explain the new syntax? I don't see the old syntax listed there either, which is perhaps an oversight?

**Comment by rjmccall** [comment]
> Does `docs/SIL.rst` need to be updated to explain the new syntax? I don't see the old syntax listed there either, which is perhaps an oversight?

Great point; I'll fix that, either in this PR if it needs a re-run or in a follow-up.

File: include/swift/AST/TypeRepr.h [hunk 1]
```diff
@@ -476,8 +476,11 @@ class FunctionTypeRepr : public TypeRepr {
   // we can have polymorphic and substituted function values.
```
Comment (varungandhi-apple) [inline include/swift/AST/TypeRepr.h]: IIUC, this comment applies to both the first and second group of fields (out of 3 groups total, with the middle one introduce in this PR), but proximity seems to indicate that it only applies to the first set of fields. Could you move things around or add an extra comment like:

```
// These fields are applicable for both AST and SIL.
```

between the second group and third group so it becomes clear the above comment also applies to the second group.

File: include/swift/AST/Types.h [hunk 2]
```diff
@@ -4065,6 +4083,10 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
     return hasResultCache() ? 2 : 0;
   }
 
+  size_t numTrailingObjects(OverloadToken<SubstitutionMap>) const {
+    return hasPatternSubstitutions() + hasInvocationSubstitutions();
```
Comment (varungandhi-apple) [inline include/swift/AST/Types.h]: It seems a bit puzzling to see two booleans being added. Could you add explicit casts here to make it clear that addition is being performed with both LHS and RHS being non-boolean integral type?

**Comment by rjmccall** [comment]
@swift-ci Please test source compatibility

**Comment by rjmccall** [comment]
Okay, full tests passed.  Made some minor changes and re-testing.

**Comment by rjmccall** [comment]
@swift-ci Please smoke test and merge.

**Comment by rjmccall** [comment]
@swift-ci Please smoke test and merge.

**Comment by rjmccall** [comment]
@swift-ci Please smoke test and merge.

**Comment by rjmccall** [comment]
@swift-ci Please smoke test and merge.

**Comment by rjmccall** [comment]
@swift-ci Please clean smoke test OS X

**Comment by rjmccall** [comment]
@swift-ci Please clean smoke test OS X platform

**Comment by jckarter** [comment]
Thanks @rjmccall! In addition to SIL.rst, it might be a good idea to mention the new syntax and generic signature behavior in the Swift Forums development thread: https://forums.swift.org/t/improving-the-representation-of-polymorphic-interfaces-in-sil-with-substituted-function-types/29711

**Comment by rjmccall** [comment]
> Thanks @rjmccall! In addition to SIL.rst, it might be a good idea to mention the new syntax and generic signature behavior in the Swift Forums development thread: https://forums.swift.org/t/improving-the-representation-of-polymorphic-interfaces-in-sil-with-substituted-function-types/29711

Great idea!  I've posted in that thread.

**Pull request merged**